### PR TITLE
IP and Interface Selection

### PR DESCRIPTION
**Pull request by cybertron**
As OpenShift is deployed in increasingly complex networking environments, we
have gotten many requests for more control over which interface is used for
the primary node IP. We provided a basic mechanism for this with
[KUBELET_NODEIP_HINT](https://github.com/openshift/machine-config-operator/pull/2888)
but as users have started to exercise that some significant limitations have
come to light.

### CODE REVIEW

**Comment by cybertron** [comment]
There are a few sections still to be filled out and I think there are some more details to be worked out, but I wanted to get this posted publicly to give the other interested parties a chance to see what direction it's headed.

File: enhancements/network/ip-interface-selection.yaml [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+## Motivation
+
+Some users want to have a great deal of control over how their network traffic
+is routed. Because we use the default route for interface and IP selection,
+in some cases they are not able to route traffic the way they want.
+
+### User Stories
+
+As a network administrator, I want cluster traffic to be on one NIC and the
```
Comment (andreaskaris) [inline enhancements/network/ip-interface-selection.yaml:41]: * As an example BZ where users might benefit from the feature: https://bugzilla.redhat.com/show_bug.cgi?id=2064131

* We have customers with multiple NICs attached to their servers and they want control over which NIC to use.

* It seems that this is a very common and important need for SNO deployments.

* I wouldn't be surprised if this was important for OpenStack deployments with manila (= 2 interfaces)

File: enhancements/network/ip-interface-selection.yaml [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+- Node IP (Kubelet and CRIO)
+- configure-ovs
+- resolv-prepender
+- Keepalived
+- Etcd
+
+In some cases (local DNS) it may not matter if the IP selected is consistent
+with the other cases, but Node IP, configure-ovs, and Keepalived all need to
+match because their functionality depends on it. I'm less familiar with the
+requirements for Etcd, but it seems likely that should match as well. In
```
Comment (andreaskaris) [inline enhancements/network/ip-interface-selection.yaml:77]: @rbbratta Had run into issues with etcd when the master nodes selected the wrong interface, so etcd is definitely some thing to consider.

File: enhancements/network/ip-interface-selection.yaml (hunk 1 above)
Comment (andreaskaris) [reply enhancements/network/ip-interface-selection.yaml:41]: @tsorya @rbbratta might help fill in the user story?
Also: https://coreos.slack.com/archives/CDCP2LA9L/p1652189014050609

File: enhancements/network/ip-interface-selection.yaml [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    there won't be any duplicate logic either.
+- nodeip-configuration runs and selects one or more IPs. It writes them to
+  the Kubelet and CRIO configuration files (this is the existing behavior).
+- nodeip-configuration also writes the following files (new behavior):
+  - /run/nodeip-configuration/primary-ip
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+  - /run/nodeip-configuration/interface
+- When configure-ovs runs, it looks for the interface file written by
+  nodeip-configuration. If found, br-ex will be created on that interface. If
```
Comment (andreaskaris) [inline enhancements/network/ip-interface-selection.yaml:111]: Question: Wouldn't configure-ovs be able to rely on the fact that nodeip-configuration always chooses an interface? That way, configure-ovs also doesn't have to deal with things such as stable interfaces selection ... all of that work would be done by nodeip-configuraiton, and that always. So, the "if not" case should never happen?

Comment (andreaskaris) [reply enhancements/network/ip-interface-selection.yaml:111]: Ok but configure-ovs still needs to be able to select an interface because of IPI deployments, correct?

File: enhancements/network/ip-interface-selection.yaml [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+  ways a user may have configured those services seems complex and error-prone.
+  Can we just require them to use this mechanism if they want custom IP
+  selection?
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
```
Comment (andreaskaris) [inline enhancements/network/ip-interface-selection.yaml:197]: We need tests for both configure-ovs as well as for this entire mechanism, ideally covering some nasty corner cases. Right now, due to the absence of tests, we are constantly at risk of introducing further issues wwhile trying to fix something.

File: enhancements/network/ip-interface-selection.yaml (hunk 3 above)
Comment (tsorya) [reply enhancements/network/ip-interface-selection.yaml:111]: After reboot nodeip-configuration will run and choose br-ex as interface but configure-ovs cleanups bridges before configuration. In that case we should not update interface file and leave old value or change configure-ovs logic

File: enhancements/network/ip-interface-selection.yaml (hunk 1 above)
Comment (tsorya) [reply enhancements/network/ip-interface-selection.yaml:41]: In SNO deployment with 2 or more nics  , we want to be sure that configure-ovs and nodeip-configuration chooses same ip, the one that user asked for, if not installation fails as router is not reachable

Comment (rbbratta) [reply enhancements/network/ip-interface-selection.yaml:41]: There was a customer request for dual-stack to separate IPv4 and IPv6 on different NICs, so we need to be aware of dual-stack issues.

File: enhancements/network/ip-interface-selection.yaml [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+## Motivation
+
+Some users want to have a great deal of control over how their network traffic
+is routed. Because we use the default route for interface and IP selection,
+in some cases they are not able to route traffic the way they want.
+
+### User Stories
+
+As a network administrator, I want cluster traffic to be on one NIC and the
+default route on a different NIC so external traffic is segregated from
```
Comment (rbbratta) [inline enhancements/network/ip-interface-selection.yaml:42]: `s/default route/default route(s)/`  There can be more that one default route, per stack, or also multi-homed with a metric.

File: enhancements/network/ip-interface-selection.yaml (hunk 2 above)
Comment (rbbratta) [reply enhancements/network/ip-interface-selection.yaml:77]: I tried to track it down once, but not an etcd expert.

https://bugzilla.redhat.com/show_bug.cgi?id=2057160#c24

cluest-etcd-operator uses `GetPreferredInternalIPAddressForNodeName` which iterates over `node.Status.Addresses`?

https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/dnshelpers/util.go#L40

```go
	for _, currAddress := range node.Status.Addresses {
		if currAddress.Type == corev1.NodeInternalIP {
			switch ipFamily {
			case "tcp4":
```

`node.Status.Addresses` is not sorted?

Is looks like kubelet `nodeAddressesFunc` provides node.Status.Addresses

https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubelet/kubelet_node_status.go#L642

```go
	// if cloud is not nil, we expect the cloud resource sync manager to exist
	var nodeAddressesFunc func() ([]v1.NodeAddress, error)
	if kl.cloud != nil {
		nodeAddressesFunc = kl.cloudResourceSyncManager.NodeAddresses
	}
```

File: enhancements/network/ip-interface-selection.yaml [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+to avoid IPs and interfaces changing. As a result, version skew should not be
+a problem.
+
+### Operational Aspects of API Extensions
+
+NA
+
+#### Failure Modes
+
+NA
```
Comment (rbbratta) [inline enhancements/network/ip-interface-selection.yaml:240]: This might not be relevant, but a common failure mode is when NetworkManager fails to activate `bond0` due to DHCP issues or mis-config and the node is presented with multiple slaves interfaces in the same subnet each with their own DHCP lease and the node has multiple IPs for the first time.

There is also the case during `fail_over_mac=1` where the `bond0` MAC will change and potentially get a different DHCP lease.

These are both maybe DHCP network design issues.

File: enhancements/network/ip-interface-selection.yaml (hunk 3 above)
Comment (tsorya) [reply enhancements/network/ip-interface-selection.yaml:111]: As a first drop  i think we can use this logic only in case node-ip-hint was set. This will unblock SNO and will give sometime to get used to this new logic without breaking anything else. WDYT?

File: enhancements/network/ip-interface-selection.yaml [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+  - /run/nodeip-configuration/interface
+- When configure-ovs runs, it looks for the interface file written by
+  nodeip-configuration. If found, br-ex will be created on that interface. If
+  not, the existing logic will be used.
+- When keepalived runs it will read the IP from the primary-ip file and the
+  interface from the interface file instead of [re-running the selection
+  logic](https://github.com/openshift/baremetal-runtimecfg/blob/62d6b95649882178ebc183ba4e1eecdb4cad7289/pkg/config/net.go#L11)
+- TODO: Etcd? I believe it has a copy of the IP selection logic from runtimecfg
```
Comment (tsorya) [inline enhancements/network/ip-interface-selection.yaml:116]: as far as i know etcd will generate certificates according to kubelet-ip

File: enhancements/network/ip-interface-selection.yaml (hunk 3 above)
Comment (tsorya) [reply enhancements/network/ip-interface-selection.yaml:111]: On the other hand, in regular BM installation we will have the same problem. kubelet ip will be set according to vip but ovs script will set something else.

File: enhancements/network/ip-interface-selection.yaml [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    as it is today. The prepender script uses the runtimecfg node-ip logic,
+    so although it won't be able to consume the output of nodeip-configuration
+    there won't be any duplicate logic either.
+- nodeip-configuration runs and selects one or more IPs. It writes them to
+  the Kubelet and CRIO configuration files (this is the existing behavior).
+- nodeip-configuration also writes the following files (new behavior):
+  - /run/nodeip-configuration/primary-ip
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+  - /run/nodeip-configuration/interface
```
Comment (tsorya) [inline enhancements/network/ip-interface-selection.yaml:109]: Saving files in /run will create issue with ovs-configuration as nodeip will find br-ex as interface and ovs will cleanup everything as a first step and only then will try to reconfigure. Maybe we should save those files in some other place and not to override interface file if br-ex was found? or we will need to split ovs logic and cleanup before nodeip?

Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:109]: That's a good point, and now I'm really wondering if we should skip the interface part of this entirely and have it work solely on IPs. It's also a problem on first boot when the IP is still on the underlying interface. After configure-ovs runs the interface here will no longer be valid because the IP will have moved to br-ex.

File: enhancements/network/ip-interface-selection.yaml (hunk 5 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:42]: Right, this should probably say _primary_ default route, because for the purposes of IP selection we only care about the highest priority default route. We can only select one IP/interface so the others are ignored.

I'm not at all certain how that additionalNetworkIDs feature works today. We don't support deployment across multiple subnets so all of your masters have to be in the same network. That's part of why I'm focused on UPI in this document. We don't support overriding the IP selection logic for IPI anyway, it's always based on the VIP.

File: enhancements/network/ip-interface-selection.yaml (hunk 3 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:111]: I think I prefer keeping the logic in configure-ovs simpler over limiting the impact of this change. I also probably need to re-title this since, as you noted, IPI is affected even if it can't use KUBELET_NODEIP_HINT.

File: enhancements/network/ip-interface-selection.yaml (hunk 1 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:41]: Hmm, separating dual stack IPs may be out of scope for this iteration, but we should keep it in mind. That's probably another argument in favor of having this work based solely on IP and not have the interface file at all.

**Comment by cybertron** [comment]
Okay, thanks for all of the discussion! Based on that and my own musings, I think we need to drop the interface from this and just persist the selected IP address(es). That way if an IP gets moved we don't have to worry about stale data, and the IP isn't allowed to change on the fly (or at all) so we don't have to worry about that.

Interrupts willing, I will try to get a new version of this pushed tomorrow that addresses all/most of the comments so far.

**Comment by tsorya** [comment]
if we want to use just ip i created 2 drafts for this, maybe it will help
https://github.com/openshift/machine-config-operator/pull/3233
https://github.com/openshift/baremetal-runtimecfg/pull/185/files

this works on ipv4 and allows to set any ip i want even if it doesn't have default route

**Comment by tsorya** [comment]
Though it opens one more question for ovn team, why we should set low metric for br-ex route that it is created. As a user i would prefer , if i know what i want, to leave same metrics we had before creating br-ex

File: enhancements/network/ip-interface-selection.yaml (hunk 7 above)
Comment (flaper87) [reply enhancements/network/ip-interface-selection.yaml:116]: > as far as i know etcd will generate certificates according to kubelet-ip

I believe this is correct

File: enhancements/network/ip-interface-selection.yaml (hunk 6 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:240]: Yeah, I think that's kind of tangential. If the IP changes out from under us that's a problem regardless of how we're selecting the IP.

**Comment by cybertron** [comment]
> if we want to use just ip i created 2 drafts for this, maybe it will help [openshift/machine-config-operator#3233](https://github.com/openshift/machine-config-operator/pull/3233) https://github.com/openshift/baremetal-runtimecfg/pull/185/files
> 
> this works on ipv4 and allows to set any ip i want even if it doesn't have default route
Yep, thanks for proposing those. I anticipate using those patches for the implementation of the design we agree on here.

> Though it opens one more question for ovn team, why we should set low metric for br-ex route that it is created. As a user i would prefer , if i know what i want, to leave same metrics we had before creating br-ex
Someone from SDN can correct me if I'm wrong, but I think it was to avoid conflicts with other default routes on the cluster. That might be something we need to revisit as part of supporting OVNK on an isolated interface.

**Comment by cybertron** [comment]
Okay, I've filled out the rest of the sections (although there are still one or two open questions) and tried to address all of the comments so far. Let me know what you think.

**Comment by tsorya** [comment]
looks good as for me

File: enhancements/network/ip-interface-selection.yaml [hunk 9]
```diff
@@ -0,0 +1,278 @@
+---
+title: ip-interface-selection
+authors:
+  - @cybertron
+reviewers:
+  - @jcaamano
+  - @tsorya
```
Comment (flaper87) [inline enhancements/network/ip-interface-selection.yaml:7]: in case you want/can/need add me as reviewer too :)

**Review by flaper87 — COMMENTED** [review]
LGTM

File: enhancements/network/ip-interface-selection.yaml [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+- When keepalived runs it will read the IP from the primary-ip file and use
+  the associated interface for VRRP traffic.
+  - Note: When [dual stack VIP support](https://github.com/openshift/enhancements/blob/master/enhancements/network/on-prem-dual-stack-vips.md)
+    is implemented it will need to look at the appropriate ipvX file for each
+    VIP. Until then, it should always use the primary IP.
+
+#### Variation [optional]
+
+We may want to rename KUBELET_NODEIP_HINT to reflect the fact that it will now
+affect more than just Kubelet.
```
Comment (maiqueb) [inline enhancements/network/ip-interface-selection.yaml:115]: I agree. The variable name should be updated to reflect its new "capabilities".

File: enhancements/network/ip-interface-selection.yaml [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+  - Answer: Yes. In IPI deployments the VIP is always used for node ip
+    selection. In an environment where the VIP subnet is isolated this could
+    result in the node IP and the interface selected by configure-ovs not
+    matching.
+
+- In UPI deployments it is also possible to set the Node IP by manually writing
+  configuration files for Kubelet and CRIO. Trying to look for all possible
+  ways a user may have configured those services seems complex and error-prone.
+  Can we just require them to use this mechanism if they want custom IP
+  selection?
```
Comment (maiqueb) [inline enhancements/network/ip-interface-selection.yaml:194]: we're waiting for an answer to this question, right ? 

Personally, it makes sense to me, but I see this as part of unifying IP and/or interface selection.

**Review by maiqueb — COMMENTED** [review]
Is `.yaml` the correct extension for this document ? Shouldn't it be a markdown file ?

File: enhancements/network/ip-interface-selection.yaml (hunk 7 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml]: Okay, I'll make a note of that in the next revision.

File: enhancements/network/ip-interface-selection.yaml (hunk 9 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:7]: Done, thanks!

File: enhancements/network/ip-interface-selection.yaml (hunk 10 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:115]: Added this to open questions to highlight that we need to agree on a name.

File: enhancements/network/ip-interface-selection.yaml (hunk 11 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.yaml:194]: I guess we could proceed without making this statement, but I think we'll continue to have problems when people use other mechanisms to set the node ip.

**Comment by cybertron** [comment]
> Is `.yaml` the correct extension for this document ? Shouldn't it be a markdown file ?

Oops, you're right. I'll fix that.

I added Dan and Tim as approvers since I think they're the most familiar with this. Also linked the epic we have for this work.

There are two remaining open questions that we should probably answer. I briefly commented about them elsewhere, but please let me know what you think and pull in anyone you think might have a vested interest in them.

File: enhancements/network/ip-interface-selection.md [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+## Implementation History
+
+KUBELET_NODEIP_HINT was implemented in 4.11 and backported to older releases
+to improve the user experience where they needed to override the default logic
+in nodeip-configuration.
+
+## Alternatives
+
+Leave things basically as they are, but teach services like configure-ovs and
```
Comment (jcaamano) [inline enhancements/network/ip-interface-selection.md:278]: Right now user should be able to configure the network how they want, although rather undocumented. It involves:

* Deploying their own keyfiles configuring br-ex on their own and using different filenames than what configure-ovs would use
* In 4.10 release or later, disabling configure-ovs (the service itself, or touching /run/configure-ovs-boot-done)

I guess one alternative would be to refine & document this.

File: enhancements/network/ip-interface-selection.md [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+Some users want to have a great deal of control over how their network traffic
+is routed. Because we use the default route for interface and IP selection,
+in some cases they are not able to route traffic the way they want.
+
+### User Stories
+
+As a deployer, I want my cluster traffic to stay on an isolated network with
+no external gateway. External traffic will travel on a different interface
+that is managed by a service like MetalLB.
```
Comment (jcaamano) [inline enhancements/network/ip-interface-selection.md:45]: Would there be this other possibility?
* running kubelet on one interface
* running the pod network on a different interface

I don't know if if it is reasonable at all, probably does not make a lot of sense. But in that it does we would not want to use nodeip-configuration decision I guess.

Comment (cybertron) [reply enhancements/network/ip-interface-selection.md:45]: Yeah, my thought is that for the purposes of this enhancement we make sure everything is consistently picking the same interface, then as a followup we come up with a more complete solution that allows deployers to pick what they want to run where. I think that's going to be a lot harder to design so I think it's worth doing it in two steps.

File: enhancements/network/ip-interface-selection.md (hunk 12 above)
Comment (cybertron) [reply enhancements/network/ip-interface-selection.md:278]: Yeah, I guess that in combination with KUBELET_NODEIP_HINT might handle a number of the use cases, but I feel like manual br-ex creation is a big hammer for a relatively small nail. I can add that to the alternatives though in case we decide to pursue that instead.

**Comment by tsorya** [comment]
It becomes more and more critical change for us as currently nvidia is stuck on the same issue and will be very glad to push it

**Comment by jcaamano** [comment]
>I certainly agree with this feeling. Especially because, as it is now, it overrides user's network configurations. This is counter intuitive when, for example, and nmstate config is passed through assisted service and the resulting network configuration doesn't match the desired topology.

@flaper87 can you explain this a bit more? How would it override the user's network configuration if manual br-ex configuration **would be** the user's network configuration, one and the same?

**Comment by flaper87** [comment]
> > I certainly agree with this feeling. Especially because, as it is now, it overrides user's network configurations. This is counter intuitive when, for example, and nmstate config is passed through assisted service and the resulting network configuration doesn't match the desired topology.
> 
> @flaper87 can you explain this a bit more? How would it override the user's network configuration if manual br-ex configuration **would be** the user's network configuration, one and the same?

Oh, I think I misread your original message. I understood you were suggesting to just let the ovs script do what it does. I see now that you are saying that users, today, can create their own `br-ex` bridge. Disregard what I said as far as "overriding the user configs" go. 

I do agree with @cybertron that it is a bit of a big hammer, tho. I see this approach as useful for much more complex network configurations.

**Comment by flaper87** [comment]
Actually, @jcaamano, following up on your comment:

> - Deploying their own keyfiles configuring br-ex on their own and using different filenames than what configure-ovs would use
> - In 4.10 release or later, disabling configure-ovs (the service itself, or touching /run/configure-ovs-boot-done) 

Is the above even supported? As in, if a user disables the service and/or configures the `br-ex` interface manually, would that cluster/configuration be considered as supported?

**Comment by jcaamano** [comment]
> Actually, @jcaamano, following up on your comment:
> 
> > * Deploying their own keyfiles configuring br-ex on their own and using different filenames than what configure-ovs would use
> > * In 4.10 release or later, disabling configure-ovs (the service itself, or touching /run/configure-ovs-boot-done)
> 
> Is the above even supported? As in, if a user disables the service and/or configures the `br-ex` interface manually, would that cluster/configuration be considered as supported?

It should work but it is not documented and not being verified so I wouldn't say it is supported. I don't think that listing alternatives in this enhancement mean that they are supported me but correct me if I am wrong.

**Comment by flaper87** [comment]
> > Actually, @jcaamano, following up on your comment:
> > > * Deploying their own keyfiles configuring br-ex on their own and using different filenames than what configure-ovs would use
> > > * In 4.10 release or later, disabling configure-ovs (the service itself, or touching /run/configure-ovs-boot-done)
> > 
> > 
> > Is the above even supported? As in, if a user disables the service and/or configures the `br-ex` interface manually, would that cluster/configuration be considered as supported?
> 
> It should work but it is not documented and not being verified so I wouldn't say it is supported. I don't think that listing alternatives in this enhancement mean that they are supported but correct me if I am wrong.

you are not wrong. That was mostly for my own curiosity and learning since I have been dealing with this issue quite a lot lately and knowing all of the alternatives definitely helps approach this in the best way possible.

@cybertron may be worth listing these alternatives and maybe we can try to get this PR through the finish line. what do you think?

**Comment by cybertron** [comment]
Oops, forgot to add the alternative we discussed earlier. New version coming shortly with the remaining open questions answered (let me know if you disagree with my answers though) and the new alternative. I think with all that taken care of we should be good to proceed with this.

**Comment by flaper87** [comment]
@cybertron markdown linter doesn't seem to agree with you :D 

Otherwise, lgtm

**Comment by cybertron** [comment]
> @cybertron markdown linter doesn't seem to agree with you :D

Turns out you need to quote the usernames or the linter complains. This one should pass (or at least it does locally).
